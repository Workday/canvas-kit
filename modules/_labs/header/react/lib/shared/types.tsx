export enum HeaderTheme {
  White,
  Blue,
  Transparent,
}

export enum HeaderVariant {
  Dub,
  Full,
  Global,
}

export enum HeaderHeight {
  Small = '64px',
  Large = '80px',
}

export type BreakpointType = 'sm' | 'md' | 'lg';
