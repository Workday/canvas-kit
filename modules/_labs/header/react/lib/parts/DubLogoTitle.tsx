import * as React from 'react';
import styled from '@emotion/styled';
import {colors, spacing} from '@workday/canvas-kit-react-core';
import {HeaderHeight, HeaderTheme} from '../shared/types';
import chroma from 'chroma-js';
import {dubLogoWhite, dubLogoBlue} from '@workday/canvas-kit-react-common'; // TODO: Replace with real SVGs from @workday/brand-assets

export type DubTitleProps = {
  /**
   * A HeaderTheme enum indicating which theme to use (White, Blue or Transparent)
   */
  themeColor: HeaderTheme;
  /**
   * The title to display in the header. Not used if `brand` is provided
   */
  title?: string;
  /**
   * The background color/gradient to be used behind the logo
   */
  bgColor?: string;
};

const LockupContainer = styled('div')({
  display: 'inline-block',
});

const Lockup = styled('div')<DubTitleProps>(
  {
    display: 'flex',
    alignItems: 'center',
    height: HeaderHeight.Small,
    paddingLeft: spacing.m,
  },
  ({bgColor}) => ({
    background: bgColor ? bgColor : 'none',
  })
);

const Title = styled('h3')<DubTitleProps>(
  {
    fontSize: '20px',
    fontWeight: 400,
    padding: `${spacing.xxs} ${spacing.s}`,
    paddingRight: spacing.l,
    marginLeft: spacing.s,
    'white-space': 'nowrap',
  },
  ({themeColor}) => ({
    color: themeColor === HeaderTheme.White ? colors.blueberry500 : colors.frenchVanilla100,
    borderLeft: `1px solid ${
      themeColor === HeaderTheme.White
        ? colors.soap400
        : chroma(colors.frenchVanilla100)
            .alpha(0.3)
            .css()
    }`,
  })
);

const DubLogo = styled('div')<DubTitleProps>({
  lineHeight: 0,
});

export class DubLogoTitle extends React.Component<DubTitleProps> {
  static defaultProps = {
    themeColor: HeaderTheme.White,
  };

  render() {
    return (
      <LockupContainer>
        <Lockup {...this.props}>
          <DubLogo
            {...this.props}
            dangerouslySetInnerHTML={{
              __html: this.props.themeColor === HeaderTheme.White ? dubLogoBlue : dubLogoWhite,
            }}
          />
          {this.props.title && <Title {...this.props}>{this.props.title}</Title>}
        </Lockup>
      </LockupContainer>
    );
  }
}
