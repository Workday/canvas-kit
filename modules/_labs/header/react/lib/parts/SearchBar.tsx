import * as React from 'react';
import styled from '@emotion/styled';
import {CSSObject} from '@emotion/core';
import {type, colors, spacing, spacingNumbers} from '@workday/canvas-kit-react-core';
import {GrowthBehavior} from '@workday/canvas-kit-react-common';
import {IconButton, IconButtonVariant} from '@workday/canvas-kit-react-button';
import {searchIcon, xIcon} from '@workday/canvas-system-icons-web';
import {FormField, FormFieldLabelPosition} from '@workday/canvas-kit-react-form-field';
import {Combobox} from '@workday/canvas-kit-labs-react-combobox';
import {TextInput} from '@workday/canvas-kit-react-text-input';
import {MenuItemProps} from '@workday/canvas-kit-labs-react-menu';
import {SearchThemeAttributes, searchThemes} from '../shared/themes';
import {SearchTheme} from '../shared/types';
import chroma from 'chroma-js';
import uuid from 'uuid/v4';

export interface SearchBarProps extends GrowthBehavior, React.FormHTMLAttributes<HTMLFormElement> {
  /**
   * The function called when the SearchBar form is submitted. The current input value is passed to the callback function.
   */
  onSubmit: React.FormEventHandler<HTMLFormElement>;
  /**
   * If true, collapse the SearchBar text input into a toggle icon. Useful for responsive layouts.
   * @default false
   */
  isCollapsed?: boolean;
  /**
   * The function called when the SearchBar text input changes.
   */
  onInputChange?: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * The autocomplete items of the SearchBar. This array of menu items is shown under the search bar.
   */
  autocompleteItems?: React.ReactElement<MenuItemProps>[];
  /**
   * The theme of the header the search input is being rendered in.
   */
  searchTheme?: SearchTheme | SearchThemeAttributes;
  /**
   * The placeholder text of the SearchBar text input.
   * @default Search
   */
  placeholder?: string;
  /**
   * The initial value of the SearchBar text input.
   */
  initialValue?: string;
  /**
   * If true, right-align the SearchBar. If false, the text input should grow to left-align the SearchBar.
   * @default false
   */
  rightAlign?: boolean;
  /**
   * The screenreader label text for the SearchBar text input.
   * @default Search
   */
  inputLabel?: string;
  /**
   * The screenreader label text for the SearchBar submit button.
   * @default Search
   */
  submitLabel: string;
  /**
   * The screenreader label text for the SearchBar clear button.
   * @default Reset Search Form
   */
  clearButtonAriaLabel?: string;
  /**
   * The screenreader label text for the button to open the collapsed SearchBar.
   * @default Open Search
   */
  openButtonLabel: string;
  /**
   * The screenreader label text for the button to close the open SearchBar.
   * @default Cancel
   */
  closeButtonLabel: string;
  /**
   * If true, render the SearchBar with a button to clear the text input.
   * @default true
   */
  showClearButton?: boolean;
}

export interface SearchBarState {
  showForm: boolean;
  searchQuery: string;
  isFocused: boolean;
}

function getInputColors(theme: SearchThemeAttributes, isFocused?: boolean) {
  return {
    background: isFocused && theme.backgroundFocus ? theme.backgroundFocus : theme.background,
    backgroundHover: theme.backgroundHover,
    color: isFocused && theme.colorFocus ? theme.colorFocus : theme.color,
    placeholderColor:
      isFocused && theme.placeholderColorFocus
        ? theme.placeholderColorFocus
        : theme.placeholderColor,
    boxShadow: isFocused && theme.boxShadowFocus ? theme.boxShadowFocus : theme.boxShadow,
  };
}

const formCollapsedBackground = colors.frenchVanilla100;

const maxWidth = 480;
const minWidth = 120;
const height = 44;

const SearchForm = styled('form')<
  Pick<SearchBarProps, 'isCollapsed' | 'rightAlign' | 'grow'> & Pick<SearchBarState, 'showForm'>
>(
  {
    position: 'relative',
    flex: `1 1 auto`, // Instead of just flex-grow: 1 for IE11, see https://github.com/philipwalton/flexbugs#flexbug-1
    display: 'flex',
    alignItems: 'center',
    marginLeft: spacing.m,
    minWidth: minWidth,
  },
  ({isCollapsed, showForm, rightAlign, grow}) => {
    const collapseStyles: CSSObject = isCollapsed
      ? {
          top: 0,
          right: 0,
          left: 0,
          bottom: 0,
          margin: 0,
          position: showForm ? 'absolute' : 'relative',
          backgroundColor: showForm ? formCollapsedBackground : 'rgba(0, 0, 0, 0)',
          transition: 'background-color 120ms',
          maxWidth: showForm ? 'none' : spacingNumbers.xl + spacingNumbers.xxs,
          minWidth: spacingNumbers.xl + spacingNumbers.xs,
          overflow: showForm ? 'visible' : 'hidden',
          zIndex: 1,
        }
      : {};
    const rightAlignStyles: CSSObject = rightAlign
      ? {
          textAlign: 'right',
          maxWidth: grow ? '100%' : maxWidth,
        }
      : {};
    return {...rightAlignStyles, ...collapseStyles};
  }
);

const SearchContainer = styled('div')({
  position: `relative`,
  width: `100%`,
  textAlign: 'left',
  minHeight: height,
  height: height, // Needed to keep IE11 vertically centered
});

const SearchCombobox = styled(Combobox)({
  width: `100%`,
});

const SearchIcon = styled(IconButton)<Pick<SearchBarProps, 'isCollapsed'> & {isHidden: boolean}>(
  ({isCollapsed, isHidden}) => {
    const collapseStyles: CSSObject = isCollapsed
      ? {
          width: spacing.xl,
          height: spacing.xl,
        }
      : {
          width: spacing.l,
          height: spacing.l,
        };

    return {
      position: `absolute`,
      margin: `auto ${spacing.xxs}`,
      top: spacing.zero,
      bottom: spacing.zero,
      zIndex: 3,
      display: isHidden ? 'none' : 'flex',
      ...collapseStyles,
    };
  }
);

const CloseButton = styled(IconButton)<
  Pick<SearchBarProps, 'isCollapsed'> & Pick<SearchBarState, 'showForm'>
>(({isCollapsed, showForm}) => {
  const collapseStyles: CSSObject =
    isCollapsed && showForm
      ? {
          display: 'inline-block',
        }
      : {
          display: 'none',
        };

  return {
    position: `absolute`,
    top: spacing.zero,
    right: spacing.zero,
    bottom: spacing.zero,
    margin: `auto ${spacing.xxs}`,
    zIndex: 3,
    ...collapseStyles,
  };
});

const SearchField = styled(FormField)<
  Pick<SearchBarProps, 'isCollapsed' | 'grow'> & Pick<SearchBarState, 'showForm'>
>(({isCollapsed, showForm, grow}) => {
  return {
    display: (isCollapsed && showForm) || !isCollapsed ? 'inline-block' : 'none',
    width: '100%',
    height: height,
    maxWidth: isCollapsed || grow ? '100%' : maxWidth,
    marginBottom: spacingNumbers.zero,
    '> div': {
      display: 'block',
    },
  };
});

const SearchInput = styled(TextInput)<
  Pick<SearchBarProps, 'isCollapsed' | 'grow'> & {inputColors: ReturnType<typeof getInputColors>}
>(({isCollapsed, inputColors, grow}) => {
  const collapseStyles: CSSObject = isCollapsed
    ? {
        ...type.h3,
        fontWeight: 400,
        paddingLeft: spacingNumbers.xl + spacingNumbers.s,
        paddingRight: spacingNumbers.xl + spacingNumbers.s,
        maxWidth: 'none',
        minWidth: 0,
        backgroundColor: `rgba(0, 0, 0, 0)`,
      }
    : {
        maxWidth: grow ? '100%' : maxWidth,
        minWidth: minWidth,
        paddingLeft: spacingNumbers.xl + spacingNumbers.xxs,
        paddingRight: spacing.xl,
        backgroundColor: inputColors.background,
      };
  return {
    ...type.body,
    boxShadow: inputColors.boxShadow,
    color: inputColors.color,
    border: 'none',
    WebkitAppearance: 'none',
    transition: 'background-color 120ms, color 120ms, box-shadow 200ms, border-color 200ms',
    zIndex: 2,
    height: height,
    paddingTop: spacing.xs,
    paddingBottom: spacing.xs,
    width: '100%',
    '&::-webkit-search-cancel-button': {
      display: 'none',
    },
    '&::placeholder': {
      color: inputColors.placeholderColor,
    },
    '&:not([disabled])': {
      '&:focus, &:active': {
        outline: 'none',
        boxShadow: inputColors.boxShadow,
      },
      '&:hover': {
        backgroundColor: inputColors.backgroundHover,
      },
    },
    ...collapseStyles,
  };
});

export class SearchBar extends React.Component<SearchBarProps, SearchBarState> {
  static Theme = SearchTheme;
  static defaultProps = {
    placeholder: 'Search',
    inputLabel: 'Search',
    submitLabel: 'Search',
    openButtonLabel: 'Open Search',
    closeButtonLabel: 'Cancel',
    showClearButton: true,
  };

  private inputRef = React.createRef<HTMLInputElement>();
  private openRef = React.createRef<HTMLButtonElement>();
  private labelId = uuid();

  state: Readonly<SearchBarState> = {
    showForm: false,
    searchQuery: '',
    isFocused: false,
  };

  private getTheme = () => {
    let theme: SearchThemeAttributes;
    if (typeof this.props.searchTheme === 'number') {
      theme = searchThemes[this.props.searchTheme];
    } else if (this.props.searchTheme) {
      theme = this.props.searchTheme;
    } else {
      theme = searchThemes[SearchTheme.Light];
    }
    return theme;
  };

  private getThemeColors = (): ReturnType<typeof getInputColors> => {
    const theme =
      this.props.isCollapsed && this.state.showForm
        ? searchThemes[SearchTheme.Transparent]
        : this.getTheme();
    return getInputColors(theme, this.state.isFocused);
  };

  getIconButtonType = (): IconButtonVariant => {
    let background = this.getThemeColors().background || `#fff`;
    if (this.props.isCollapsed && this.state.showForm) {
      background = formCollapsedBackground;
    }
    const isDarkBackground = chroma(background).get('lab.l') < 70;
    return isDarkBackground ? IconButton.Variant.Inverse : IconButton.Variant.Plain;
  };

  handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    if (this.state.searchQuery.trim()) {
      this.props.onSubmit(event);
    } else {
      this.focusInput();
    }
  };

  openCollapsedSearch = (): void => {
    if (this.props.isCollapsed && !this.state.showForm) {
      this.setState({showForm: true});
    }
  };

  closeCollapsedSearch = (): void => {
    if (this.props.isCollapsed && this.state.showForm) {
      this.setState({showForm: false});
    }
  };

  componentDidUpdate(prevProps: SearchBarProps, prevState: SearchBarState) {
    const showFormToggled = this.state.showForm !== prevState.showForm;
    if (showFormToggled) {
      if (this.state.showForm) {
        this.focusInput();
      } else {
        this.focusOpen();
      }
    }
  }

  focusInput = (): void => {
    if (this.inputRef.current) {
      this.inputRef.current.focus();
    }
  };

  focusOpen = (): void => {
    if (this.openRef.current) {
      this.openRef.current.focus();
    }
  };

  handleFocus = (): void => {
    this.setState({isFocused: true});
  };

  handleBlur = (): void => {
    this.setState({isFocused: false});
  };

  handleSearchInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    event.preventDefault();
    this.setState({searchQuery: event.target.value});
    if (this.props.onInputChange) {
      this.props.onInputChange(event);
    }
  };

  render() {
    const {
      grow,
      onSubmit,
      isCollapsed,
      onInputChange,
      autocompleteItems,
      initialValue,
      searchTheme,
      placeholder,
      rightAlign,
      inputLabel,
      submitLabel,
      showClearButton,
      clearButtonAriaLabel,
      closeButtonLabel,
      openButtonLabel,
      ...elemProps
    } = this.props;

    return (
      <SearchForm
        role="search"
        action="."
        rightAlign={rightAlign}
        grow={grow}
        aria-labelledby={this.labelId}
        isCollapsed={isCollapsed}
        onSubmit={this.handleSubmit}
        showForm={this.state.showForm}
        {...elemProps}
      >
        <SearchContainer>
          <SearchIcon
            aria-label={submitLabel}
            icon={searchIcon}
            isCollapsed={isCollapsed}
            variant={this.getIconButtonType()}
            type="submit"
            isHidden={!!isCollapsed && !this.state.showForm}
          />
          <SearchIcon
            aria-label={openButtonLabel}
            icon={searchIcon}
            isCollapsed={isCollapsed}
            variant={this.getIconButtonType()}
            onClick={this.openCollapsedSearch}
            buttonRef={this.openRef}
            type="button"
            isHidden={!isCollapsed || (!!isCollapsed && this.state.showForm)}
          />
          <SearchField
            grow={grow}
            id={this.labelId}
            inputId={`input-${this.labelId}`}
            label={inputLabel}
            labelPosition={FormFieldLabelPosition.Hidden}
            useFieldset={false}
            isCollapsed={isCollapsed}
            showForm={this.state.showForm}
          >
            <SearchCombobox
              initialValue={initialValue}
              clearButtonVariant={this.getIconButtonType()}
              autocompleteItems={autocompleteItems}
              onChange={this.handleSearchInputChange}
              onFocus={this.handleFocus}
              onBlur={this.handleBlur}
              showClearButton={!isCollapsed && showClearButton}
              clearButtonAriaLabel={clearButtonAriaLabel || 'Reset Search Form'}
              labelId={this.labelId}
            >
              <SearchInput
                inputRef={this.inputRef}
                value={this.state.searchQuery}
                placeholder={placeholder}
                isCollapsed={isCollapsed}
                inputColors={this.getThemeColors()}
                name="search"
              />
            </SearchCombobox>
          </SearchField>
          <CloseButton
            aria-label={closeButtonLabel}
            icon={xIcon}
            isCollapsed={isCollapsed}
            variant={IconButton.Variant.Plain}
            showForm={this.state.showForm}
            onClick={this.closeCollapsedSearch}
            type="button"
          />
        </SearchContainer>
      </SearchForm>
    );
  }
}
