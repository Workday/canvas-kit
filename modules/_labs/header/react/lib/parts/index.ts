export {DubLogoTitle} from './DubLogoTitle';
export {Search} from './Search';
export {WorkdayLogoTitle} from './WorkdayLogoTitle';
