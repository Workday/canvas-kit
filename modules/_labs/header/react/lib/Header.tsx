import * as React from 'react';
import {css} from '@emotion/core';
import styled from '@emotion/styled';
import {borderRadius, spacing, type} from '@workday/canvas-kit-react-core';
import {DubLogoTitle, WorkdayLogoTitle} from './parts';
import {themes} from './shared/themes';
import {HeaderHeight, HeaderTheme, HeaderVariant} from './shared/types';
import {IconButton, IconButtonProps} from '@workday/canvas-kit-react-button';
import {SystemIcon, SystemIconProps} from '@workday/canvas-kit-react-icon';
import {justifyIcon} from '@workday/canvas-system-icons-web';
import {PickRequired} from '@workday/canvas-kit-react-common';

export interface HeaderProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * The custom menu toggle node of the Header. This React node replaces the default menu toggle.
   */
  menuToggle?: React.ReactNode;
  /**
   * The theme of the Header. Accepts `White`, `Blue`, or `Transparent`.
   * @default HeaderTheme.White
   */
  themeColor?: HeaderTheme;
  /**
   * The variant of the Header. Accepts `Dub` (small) or `Full` (large).
   * @default HeaderVariant.Dub
   */
  variant?: HeaderVariant;
  /**
   * The text of the Header title. Not used if `brand` is provided.
   */
  title?: string;
  /**
   * The custom brand node of the Header. This React node replaces the dub logo and title.
   */
  brand?: React.ReactNode;
  /**
   * The url of the Header logo link.
   */
  brandUrl?: string;
  /**
   * If true, center the Header navigation. If false, right-align the Header navigation.
   * @default false
   */
  centeredNav?: boolean;
  /**
   * The function called when the responsive menu icon is clicked.
   */
  onMenuClick?: (e: React.MouseEvent) => void;
  /**
   * The React element to render in the left slot of the Header. This is typically a SearchBar component.
   */
  leftSlot?: React.ReactElement;
  /**
   * If true, render the Header in collapsed mode.
   * @default false
   */
  isCollapsed?: boolean;
}

const childrenSpacing = spacing.s;

const HeaderShell = styled('div')<PickRequired<HeaderProps, 'themeColor', 'variant'>>(
  {
    display: 'flex',
    alignItems: 'center',
    boxSizing: 'border-box',
    ...type.body,
    WebkitFontSmoothing: 'antialiased',
    MozOsxFontSmoothing: 'grayscale',
    position: 'relative',
  },
  ({variant, themeColor}) => ({
    // Only the variant Full has a large header, all the other one (Dub, Global) have a small header height
    height: variant === HeaderVariant.Full ? HeaderHeight.Large : HeaderHeight.Small,
    background: themes[themeColor].background,
    ...themes[themeColor].depth,
    color: themes[themeColor].color,
  })
);

const BrandSlot = styled('div')(
  {
    height: '100%',
  },
  (props: {grow?: boolean}) => ({
    flex: props.grow ? `1 0 auto` : 'unset',
  })
);

const BrandLink = styled('a')({
  ...type.link,

  '&:hover': {
    textDecoration: 'none',
  },

  '&:focus': {
    textDecoration: 'none',
  },
});

const navStyle = ({themeColor}: PickRequired<HeaderProps, 'themeColor', 'css'>) => {
  const theme = themes[themeColor];

  return css({
    nav: {
      display: 'flex',
      flex: `1 0 auto`, // Instead of just flex-grow: 1 for IE11, see https://github.com/philipwalton/flexbugs#flexbug-1
      justifyContent: 'center',
      height: 'inherit',
      marginLeft: spacing.xl,

      '& ul': {
        color: theme.linkColor,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        listStyleType: 'none',
        padding: 0,
        margin: 0,
        height: 'inherit',
        '&:hover': {
          // De-emphasizes all color on hover, this allows us to create an illusion that hovering over a specific element
          // fades out the rest
          color: theme.linkFadeOutColor,
        },
        '& li': {
          boxSizing: 'border-box',
          position: 'relative',
          display: 'flex',
          alignItems: 'center',
          fontSize: '14px',
          fontWeight: 700,
          height: 'inherit',
          transition: `color 150ms ease-out 0s`,
          cursor: 'pointer',
          '&:first-child > *': {
            marginLeft: 0,
          },
          '&:last-child > *': {
            marginRight: 0,
          },
          '& > *': {
            color: 'inherit',
            textDecoration: 'none',
            padding: `0px ${spacing.s}`,
            margin: `0 ${spacing.xxxs}`,
            display: 'flex',
            alignItems: 'center',
            height: 'inherit',
            '&:visited': {
              color: 'inherit', // Keeps visited links from becoming default purple
            },
          },
          '&:hover, &:active': {
            color: theme.linkColor, // Completes the illusion of sibling elements fading into the background on hover
          },
          '&.current': {
            color: theme.currentLinkColor,
            '& a': {
              cursor: 'default',
            },
            '&:hover, &:active': {
              color: theme.currentLinkColor,
            },
            '&:after': {
              position: 'absolute',
              bottom: 0,
              left: 0,
              content: `''`,
              height: 4,
              width: '100%',
              backgroundColor: theme.chipColor,
              borderRadius: `${borderRadius.m} ${borderRadius.m} 0 0`,
            },
          },
        },
      },
    },
  });
};

const ChildrenSlot = styled('div')<
  PickRequired<HeaderProps, 'themeColor', 'isCollapsed' | 'centeredNav'>
>(
  {
    marginRight: spacing.m,
    // TODO: remove this when we get real icon buttons
    '> .canvas-header--menu-icon': {
      cursor: 'pointer',
    },
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    height: '100%',
    '> *': {
      marginLeft: childrenSpacing,
    },
  },
  ({centeredNav, isCollapsed}) => ({
    '> *:not(.canvas-header--menu-icon)': {
      display: isCollapsed ? 'none' : 'flex',
    },
    '> *:last-child': {
      marginRight: isCollapsed ? '' : 0,
    },
    flex: !isCollapsed && centeredNav ? `1 0 auto` : 'unset',
  }),
  navStyle
);

class Brand extends React.Component<
  Pick<HeaderProps, 'variant' | 'brand' | 'title' | 'themeColor'>
> {
  render() {
    const {variant = HeaderVariant.Dub, brand, themeColor = HeaderTheme.White, title} = this.props;

    switch (variant) {
      case HeaderVariant.Global: {
        return <span>{brand}</span>;
      }
      case HeaderVariant.Full: {
        return (
          <span>
            {brand || <WorkdayLogoTitle title={title ? title : ''} themeColor={themeColor} />}
          </span>
        );
      }
      // HeaderVariant.Dub is default
      default: {
        return (
          <span>
            {brand || <DubLogoTitle title={title ? title : ''} themeColor={themeColor} />}
          </span>
        );
      }
    }
  }
}

class MenuIconButton extends React.Component<
  Pick<HeaderProps, 'themeColor' | 'menuToggle' | 'onMenuClick'>
> {
  render() {
    const {themeColor = HeaderTheme.White, menuToggle, onMenuClick} = this.props;
    if (menuToggle) {
      const menuToggleElement = menuToggle as React.ReactElement<any>;
      const onClick = menuToggleElement.props.onClick
        ? menuToggleElement.props.onClick
        : onMenuClick;

      return React.cloneElement(menuToggleElement, {
        onClick,
        className: 'canvas-header--menu-icon',
      });
    }

    return (
      <IconButton
        variant={
          themeColor === HeaderTheme.White ? IconButton.Variant.Circle : IconButton.Variant.Inverse
        }
        icon={justifyIcon}
        className={'canvas-header--menu-icon'}
        aria-label="Open Menu"
        onClick={onMenuClick}
      />
    );
  }
}

export default class Header extends React.Component<HeaderProps, {}> {
  static Theme = HeaderTheme;
  static Variant = HeaderVariant;

  /**
   * Helper that recursively maps ReactNodes to their theme-based equivalent.
   * Any children that are included in a Header may need to undergo transformation
   * before stamped out in render().
   *
   * E.g. <SystemIcon> components need to have the appropriate `color` and `colorHover`
   * props set based on the theme.
   *
   * @param children From props.children of a React component
   *
   * @returns {React.ReactNode} The child/children to be rendered
   */
  private renderChildren(children: React.ReactNode): React.ReactNode {
    return React.Children.map(children, child => {
      if (!React.isValidElement(child)) {
        return child;
      }

      // child is now guaranteed to be a valid ReactElement from the check above

      type Props = {children: React.ReactNode};
      const propsChildren = (child as React.ReactElement<Props>).props.children;
      const singleChild =
        React.Children.count(propsChildren) === 1 && (propsChildren as React.ReactElement<any>);
      const iconButtonVariant =
        this.props.themeColor === HeaderTheme.White
          ? IconButton.Variant.Circle
          : IconButton.Variant.Inverse;

      // Convert old method of SystemIcon into IconButton. If SystemIcon is within a link, make sure it's passed through
      if (child.type === 'a' && singleChild && singleChild.type === SystemIcon) {
        const href = (child.props as React.AnchorHTMLAttributes<HTMLAnchorElement>).href;
        const iconButtonProps = {
          onClick: () => {
            if (href) {
              window.location.href = href;
            }
          },
          variant: iconButtonVariant,
          icon: (singleChild.props as SystemIconProps).icon,
        };

        return <IconButton {...iconButtonProps} aria-label="" />;
      }

      // If child has children, render them
      if (React.Children.count(propsChildren)) {
        return React.cloneElement(child as React.ReactElement<Props>, {
          children: this.renderChildren(propsChildren),
        });
      }

      // Convert old method of SystemIcon into IconButton
      if (child.type === SystemIcon) {
        const icon = (child.props as SystemIconProps).icon;

        return <IconButton variant={iconButtonVariant} icon={icon} aria-label={icon.name} />;
      }

      // Plain icon buttons have negative margin that we need to negate.
      if (
        child.type === IconButton &&
        (child.props as IconButtonProps).variant === IconButton.Variant.Plain
      ) {
        return React.cloneElement(child as React.ReactElement<IconButtonProps>, {
          style: {margin: `0 0 0 ${childrenSpacing}`},
        });
      }

      return child;
    });
  }

  render() {
    const {
      menuToggle,
      themeColor = HeaderTheme.White,
      variant = HeaderVariant.Dub,
      centeredNav,
      title,
      brand,
      brandUrl,
      onMenuClick,
      leftSlot,
      isCollapsed,
      children,
      ...elemProps
    } = this.props;

    /* Push everything to the right if:
       - on tablet and mobile screens
       - Search isn't enabled and the nav shouldn't be centered
       - Search is enabled, and there aren't any children
    */
    const growBrand = isCollapsed || (!leftSlot && !centeredNav) || (leftSlot && !children);

    // Ignore centeredNav if search is enabled
    const shouldCenteredNav = leftSlot ? false : centeredNav;

    return (
      <HeaderShell variant={variant} themeColor={themeColor} {...elemProps}>
        <BrandSlot grow={growBrand}>
          {brandUrl ? (
            <BrandLink href={brandUrl}>
              <Brand variant={variant} brand={brand} title={title} themeColor={themeColor} />
            </BrandLink>
          ) : (
            <Brand variant={variant} brand={brand} title={title} themeColor={themeColor} />
          )}
        </BrandSlot>
        {leftSlot}
        <ChildrenSlot
          themeColor={themeColor}
          centeredNav={shouldCenteredNav}
          isCollapsed={isCollapsed}
        >
          {isCollapsed ? (
            // Screen size is smaller than our largest breakpoint so turn nav into a hamburger
            // TODO: This needs to get changed to IconButton when we get it restyled for headers
            <MenuIconButton
              themeColor={themeColor}
              menuToggle={menuToggle}
              onMenuClick={onMenuClick}
            />
          ) : (
            this.renderChildren(children)
          )}
        </ChildrenSlot>
      </HeaderShell>
    );
  }
}
