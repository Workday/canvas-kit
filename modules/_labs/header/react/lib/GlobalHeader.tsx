import * as React from 'react';
import Header from './Header';
import {HeaderVariant} from './shared/types';
import {DubLogoTitle} from './parts';

export interface GlobalHeaderProps {
  /**
   * The custom brand node of the GlobalHeader. This React node replaces the dub logo and title.
   * @default DubLogoTitle
   */
  brand: React.ReactNode;
  /**
   * The custom menu toggle node of the GlobalHeader. This React node replaces the default menu toggle.
   */
  menuToggle?: React.ReactNode;
  /**
   * The function called when the responsive menu icon is clicked.
   */
  onMenuClick?: (e: React.SyntheticEvent) => void;
  /**
   * If true, render the GlobalHeader in collapsed mode.
   * @default false
   */
  isCollapsed?: boolean;
  /**
   * The React element to render in the left slot of the GlobalHeader. This is typically a SearchBar component.
   */
  leftSlot?: React.ReactElement;
}

export default class GlobalHeader extends React.Component<GlobalHeaderProps> {
  static defaultProps = {
    brand: <DubLogoTitle />,
  };
  public render() {
    const {
      brand,
      menuToggle,
      onMenuClick,
      isCollapsed,
      leftSlot,
      children,
      ...elemProps
    } = this.props;
    return (
      <Header
        brand={brand}
        menuToggle={menuToggle}
        leftSlot={leftSlot}
        onMenuClick={onMenuClick}
        variant={HeaderVariant.Global}
        isCollapsed={isCollapsed}
        {...elemProps}
      >
        {children}
      </Header>
    );
  }
}
