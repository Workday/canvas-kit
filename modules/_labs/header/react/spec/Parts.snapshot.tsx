import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {DubLogoTitle, WorkdayLogoTitle} from '../lib/parts';
import {HeaderTheme} from '../lib/shared/types';

describe('Header Parts Snapshots', () => {
  test('Renders a DubLogoTitle with custom background', () => {
    const tree = renderer.create(<DubLogoTitle title={'test'} bgColor={'pink'} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Renders a blue-themed WorkdayLogoTitle', () => {
    const tree = renderer
      .create(<WorkdayLogoTitle title={'test'} themeColor={HeaderTheme.Blue} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
