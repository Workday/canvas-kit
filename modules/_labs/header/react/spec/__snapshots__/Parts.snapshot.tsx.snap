// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Parts Snapshots Renders a DubLogoTitle with custom background 1`] = `
.emotion-6 {
  display: inline-block;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 64px;
  padding-left: 24px;
  background: pink;
}

.emotion-0 {
  line-height: 0;
}

.emotion-2 {
  font-size: 20px;
  font-weight: 400;
  padding: 8px 16px;
  padding-right: 32px;
  margin-left: 16px;
  white-space: nowrap;
  color: #005cb9;
  border-left: 1px solid #dfe2e6;
}

<div
  className="emotion-6 emotion-7"
>
  <div
    className="emotion-4 emotion-5"
    title="test"
  >
    <div
      className="emotion-0 emotion-1"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<img src=\\"//design.workday.com/images/ck-dub-logo-blue.svg\\">",
        }
      }
      title="test"
    />
    <h3
      className="emotion-2 emotion-3"
      title="test"
    >
      test
    </h3>
  </div>
</div>
`;

exports[`Header Parts Snapshots Renders a blue-themed WorkdayLogoTitle 1`] = `
.emotion-6 {
  display: inline-block;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 80px;
}

.emotion-0 {
  padding-left: 24px;
  line-height: 0;
}

.emotion-2 {
  font-size: 20px;
  font-weight: 400;
  margin-left: 32px;
  padding: 16px;
  padding-left: 24px;
  padding-right: 32px;
  white-space: nowrap;
  display: initial;
  color: #ffffff;
  border-left: 1px solid rgba(223,226,230,0.4);
}

<div
  className="emotion-6 emotion-7"
>
  <div
    className="emotion-4 emotion-5"
    title="test"
  >
    <span
      className="emotion-0 emotion-1"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<img src=\\"//design.workday.com/images/ck-wday-logo-white.svg\\">",
        }
      }
      title="test"
    />
    <h3
      className="emotion-2 emotion-3"
      title="test"
    >
      test
    </h3>
  </div>
</div>
`;
